quantum/rgblight.c

```diff
@@ -662,10 +666,20 @@ void rgblight_sethsv_slave(uint16_t hue, uint8_t sat, uint8_t val) {
 #ifndef RGBLIGHT_CUSTOM_DRIVER
 void rgblight_set(void) {
   if (rgblight_config.enable) {
+      LED_TYPE *ledp;
+    #ifdef RGBLIGHT_LED_MAP
+      LED_TYPE led0[RGBLED_NUM];
+      for(uint8_t i = 0; i < RGBLED_NUM; i++) {
+          led0[i] = led[led_map[i]];
```

Comment:
The remapping would probably be better in the rgblight_setrgb functions instead of this send to device function. As you then only need 1 buffer as you remap in place instead of this temp copy.

Reply:
This `led0 [RGBLED_NUM];` is reserved on the stack when `rgblight_set()` is executed. I think that there is no problem because it will be released when returning from `rgblight_set()`.

Reply:
It's not that this is a problem, but that you are doing work you just don't have to do with a slight change. Think of it this way, the LED_TYPE struct is 24 bits, on some keyboards there are easily 70 LEDs or more. That's 1,680 bits (210 bytes) you are consuming for no reason.

Reply:
I know that the size of the LED_TYPE struct is 3 bytes. I also know that it will be 210 bytes with 70 LEDs.

However, remap can not be achieved only by correcting the rgblight_setrgb_X() function group. Because rgblight.h has exported the following interfaces.

```c
extern LED_TYPE led[RGBLED_NUM];
void sethsv(uint16_t hue, uint8_t sat, uint8_t val, LED_TYPE *led1);
void setrgb(uint8_t r, uint8_t g, uint8_t b, LED_TYPE *led1);
```

Users can easily avoid `rgblight_setrgb_X()` like `rgblight_effect_snake()`. Please look at the following.

https://github.com/qmk/qmk_firmware/blob/a0c36ed25318e863d5c47cf2b80423abd934c198/quantum/rgblight.c#L873-L913

Reply:
well, shit, good point. (added to list to fix later)

Reply:
And `rgblight_setrgb_at()` always calls `rgblight_set()` internally. What it means is that if you call `rgblight_setrgb_at()` 50 times to change the color of 50 LEDs, it will do 49 wasted `rgblight_set()` calls.

Reply:
I can atest to this first hand.  It's not just wasted, but it can seriously impact performance, especially on AVR chips.